core/src/Cabana_Experimental_NeighborList.hpp

```diff
@@ -179,8 +220,59 @@ auto makeNeighborList( Tag, Slice const &coordinate_slice,
                Impl::makePredicates( coordinate_slice, first, last, radius ),
                Impl::NeighborDiscriminatorCallback<Tag>{}, indices, offset );
 
-    return CrsGraph<typename DeviceType::memory_space, Tag>{
-        std::move( indices ), std::move( offset ), first, bvh.size()};
+    return CrsGraph<MemorySpace, Tag>{std::move( indices ), std::move( offset ),
+                                      first, bvh.size()};
+}
+
+template <typename MemorySpace, typename Tag>
+struct Dense
+{
+    Kokkos::View<int *, MemorySpace> cnt;
+    Kokkos::View<int **, MemorySpace> val;
+    typename MemorySpace::size_type shift;
+    typename MemorySpace::size_type total;
+};
+
+template <typename DeviceType, typename Slice, typename Tag>
+auto make2DNeighborList( Tag, Slice const &coordinate_slice,
+                         typename Slice::size_type first,
+                         typename Slice::size_type last,
+                         typename Slice::value_type radius )
+{
+    using MemorySpace = typename DeviceType::memory_space;
+    using ExecutionSpace = typename DeviceType::execution_space;
+    ExecutionSpace space{};
+
+    ArborX::BVH<MemorySpace> bvh( space, coordinate_slice );
+
+    Kokkos::View<int *, DeviceType> indices( "indices", 0 );
+    Kokkos::View<int *, DeviceType> offset( "offset", 0 );
+
+    auto const predicates =
+        Impl::makePredicates( coordinate_slice, first, last, radius );
+
+    auto const n_queries = ArborX::Traits::
+        Access<decltype( predicates ), ArborX::Traits::PredicatesTag>::size(
+            predicates );
+
+    Kokkos::View<int *, DeviceType> counts( "counts", n_queries );
+    bvh.query(
```

Comment:
OK so you do your own 2-pass here - do you force ArborX to do the queries in 1-pass mode so this is effectively still just a 2-pass process as you are filling your own data structure?

Reply:
I guess you must check if there is enough space allocated by the user but you never fill the native data structure so ArborX will always do 1-pass because the internal count never goes past 0?

Reply:
Correct.  We detect that you never called the output functor and do a single pass.  Note that we are working on adding a new `BVH::query()` overload that does only takes the callback (no offsets and values).